NEP: Initial draft for NEP 43 for extensible ufuncs

seberg opened the pull request:
There is still quite a bit to be done here. However, some of the important things for discussion are there I think, so putting it out here to work on here and hope for feedback.

One of the thoughts I had with respect to https://github.com/numpy/numpy/issues/12518 is to pass in a struct (which could grow) with all of the important information.  The main tricky thing is to decide on:

1. Promotion (but I think multiple dispatched promotion is a good solution)
2. How to best wrap existing loops, and how to make the whole API as limited as possible for now.

seberg committed:
NEP: Initial draft for NEP 43

mhvk commented on doc/neps/nep-0043-extensible-ufuncs.rst:
```diff
@@ -0,0 +1,631 @@
+Both of these are *not* stored on the ``UFuncImpl`` itself and just like
+methods, most of the time ``UFuncImpl`` should only be used indirectly by
+calling the ufunc it is bound to.
+It is possible to bind a single ``UFuncImpl`` to multiple ufuncs.
```
I think this is crucial and deserves mention early on! Here, a specific example might be good. E.g., I think `sin` and `cos` could share complete implementations, but `add` and `subtract` could not (e.g., `add` being defined for strings but not for `datetime`, and vice versa for `substract`).

mhvk commented on doc/neps/nep-0043-extensible-ufuncs.rst:
```diff
@@ -0,0 +1,631 @@
+a better extensibility of the ufunc interface, which is the backbone for
+a large part of the NumPy API.
+Especially, parametric DTypes are currently severly limited, making it
+impossible to implement for a user-defined DType to register universal
```
delete "to implement" here.

mhvk commented on doc/neps/nep-0043-extensible-ufuncs.rst:
```diff
@@ -0,0 +1,631 @@
+to error reporting, by extending the signature of the core inner-loop
+functions.
+Finally, one of the most important steps in the ufunc call is finding the
+correct function for a specific set of input dtypes.  We propose to use
```
correct *inner loop* function? Or just "correct inner loop"?

hameerabbasi replied:
How about "correct ufunc implementation"?

mhvk commented on doc/neps/nep-0043-extensible-ufuncs.rst:
```diff
@@ -0,0 +1,631 @@
+implementation only.
+
+
+Bound UFuncImpl
```
This section ended up confusing me no end. It is not obvious to me that it is like a bound method where the ufunc and dtypes combine to some form of `self`, unless one thinks of each `UFuncImpl` as some type of class that has `__init__(self, ufunc, dtypes)`, so that all the methods can just refer to `self.ufunc` and `self.dtypes`.

Indeed, I'm sufficiently confused that it makes me wonder whether it might not be helpful to have a mock-up of a python implementation of a `UFuncImpl` and on how one would use it for an existing function (the units case would be a good one), as well as how one would write a new `UFunc` taking perhaps an existing `UFuncImpl` (say, a `sincos` implementation). (This echoes a point made by @mattip in the original thread...)

seberg replied:
Yeah, I guess you could see it that way with an `__init__(self, ufunc, dtypes)`, although, I do not think such an init has to technically be exposed to the user.  Anyway, I have to think about trying to clear this up, I was not sure I like the mental model around having an `__init__`, it somewhat implies there is a full blown (subclass-able?) class, something that I would like to avoid at least for now.

mhvk replied:
Agreed that the way it is exposed to a python user does not have to be decided now. My main comment here was really that I found the present text confusing (unlike the rest). Might it make sense to move this piece to a separate very preliminary discussion of how things might be exposed in python?

 Or, not as good but perhaps more practical, remove it and just have a note up front that a python implementation is not yet fully considered but that it is attempted to make sure nothing would stop it?

seberg replied:
I have expanded/refactored this section, and more importantly tried to add a super-highlevel-get-the-feel for things section at the start, which happens to include `resolve_impl` (I don't know how much that helps, and I guess you are the wrong person to ask by now).

In any case, this section is restructured, and I tried to point out that it is a bit nitty-gritty for the interested reader (I am unsure about it, but it helped me be clear about who has what information)

mhvk commented on doc/neps/nep-0043-extensible-ufuncs.rst:
```diff
@@ -0,0 +1,631 @@
+   inputs and tries again.  This is well defined for most mathematical
+   functions, but can be disabled or customized if necessary.
+
+2. Users can *register* new Promoters just as they can register new UFuncImpl.
```
At this point, registration of `UFuncImpl` has not really been discussed much.

mhvk commented on doc/neps/nep-0043-extensible-ufuncs.rst:
```diff
@@ -0,0 +1,631 @@
+            return NotImplemented
+
+        # Could check that res[1] is actually Int64
+        return ufunc.resolve_impl(tuple(res))
```
The mock-up is very useful, but would be even better if `resolve_impl` had already been introduced! (perhaps via the larger mock-up I wondered about above).

mhvk commented on doc/neps/nep-0043-extensible-ufuncs.rst:
```diff
@@ -0,0 +1,631 @@
+
+* How ``CastingImpl`` lookup, and thus the decision whether a cast is possible,
+  is defined. (This is speed relevant, although mainly during a transition
+  phase where UFuncs where NEP 43 is not yet implemented).
```
double where

mhvk reviewed (commented):
@seberg - before anything else, a very nice write-up! Some initial comments, some just details, some perhaps too much focussed on implementation, and one real confusion, leading to the perhaps the main suggestion of some type of python mock-up.

seberg commented on doc/neps/nep-0043-extensible-ufuncs.rst:
```diff
@@ -0,0 +1,631 @@
+  an ``resolve_descriptors`` function (similar to ``adjust_descriptors``
+  of ``CastingImpl``)::
+
+      int resolve_descriptors(
```
Oh, there is no intention for in-place, the naming is just confusing `in` and `out` was not supposed to refer to input and output descriptors...

seberg commented on doc/neps/nep-0043-extensible-ufuncs.rst:
```diff
@@ -0,0 +1,631 @@
+  setting the desired output descriptors, and possibly modifying the input
+  ones to conform with the inner-loops requirements.
+
+* Define a new structure to be passed in the inner-loop, which can be
```
Have to think about this, thanks.  Yes, we need to wrap legacy loops in any case I think (or, hidden from the user, have a flag which causes the old inner-loop convention (depending on what is easier probably, I don't mind the indirection for loops defined outside of NumPy, and loops inside of NumPy should move ASAP as well).

seberg commented on doc/neps/nep-0043-extensible-ufuncs.rst:
```diff
@@ -0,0 +1,631 @@
+   ``Unit(float64, "m")``.
+
+The difference between the two, is that in the first case a ``UFuncImpl``
+can be created which *wraps* a single existing ``UFuncImpl`` and after that
```
Ah, this is great to here! I was seriously struggling with trying to figure out a good way to make Option 2 something that does not feel out of place...

mhvk commented on doc/neps/nep-0043-extensible-ufuncs.rst:
```diff
@@ -0,0 +1,631 @@
+              PyUFuncObject *ufunc, PyArray_DTypeMeta *DTypes[nargs],
+              PyArray_Descr *in_dtypes[nargs], PyArray_Descr *out_dtypes[nargs]);
+
+  setting the desired output descriptors, and possibly modifying the input
```
@seberg - the "and possibly modifying the input" here was what made me think that changes were made in-place. Maybe rephrase this?

seberg commented on doc/neps/nep-0043-extensible-ufuncs.rst:
```diff
@@ -0,0 +1,631 @@
+
+   * Clearing of floating point exception flags (if necessary),
+   * Possibly allocating temporary working space,
+   * Setting (and potentially finding) the inner-loop function.  Finding
```
True, although I think this can be part of the setup?  This is about specializations, such as contiguous loops (SIMD capable strides), or reduction loops, which can be implemented faster.  I guess the grouping here is trying to find the most practical grouping for users who want the most control possible, while not asking them to take charge of setting up the outer iteration as well.

Anyway, thanks for this and all the feedback, will try to iterate/think on it!

mhvk replied:
Ah, with that goal it does make sense. I think then all that's needed is to give exactly that example!

anirudh2290 commented on doc/neps/nep-0043-extensible-ufuncs.rst:
```diff
@@ -0,0 +1,631 @@
+  user input descriptors with descriptors matching the wrapped loop.
+  It must be possible to *view* the inputs as the output.
+* ``wrap_outputs(Tuple[DType]: input_descr) -> Tuple[DType]`` replacing the
+  resolved descriptors with with the desired actual loop descriptors.
```
I was wondering how this handles a case like np.add(Unit[Float64]("m"), Unit[Float64["km"]) ?

seberg replied:
Thanks, good point, I should explain that/give an example. For Units this uses `.view(np.float64)` in the inner loop effectively. In the outer loop, the descriptors will be set to `(Unit[Float64]("m"), Unit[Float64["m"], Unit[Float64]("m"))`, so the casting machinery has to convert the `"km"` to `"m"`.

seberg commented on doc/neps/nep-0043-extensible-ufuncs.rst:
```diff
@@ -0,0 +1,631 @@
+  flagging (e.g. warnings) if desired.
+
+* A flag to ask NumPy to perform floating point error checking (after custom
+  setup and before user teardown).
```
Many integer loops or loops working with Python objects only, do not really have to. And apparently floating point error checking can be a bit slow on some hardware, so I think we can create such an option.  Although, I guess it would be fair to add that later as well.

seberg committed:
NEP: Small updates to NEP 43

seberg commented:
@hameerabbasi (and quite honestly everyone). While I will iterate on it in the next days, I think the big picture stuff is there and should be fine to read, so please do whenever you have time, independently of me updating it soon.  It will require more iterations in any case, delaying the start of the discussion just means that my iterations are more annoying, slower and less thorough.

seberg commented on doc/neps/nep-0043-extensible-ufuncs.rst:
```diff
@@ -0,0 +1,631 @@
+  was already set previously it will be stored. Any previous error
+  will be cleared before calling ``teardown``.
+  The return value of the inner-loop is passed in to allow simple error
+  flagging (e.g. warnings) if desired.
```
Note for those who already read this. I think I will modify this. I am starting to be convinced that `teardown` does not need to do any error-related work, and that it may be an anti-pattern.  Floating point errors are special, since they require this.  As such, we could allow an error return, but should discourage it, since its much like deleting a Python object.

seberg committed:
Partial iteration on some content, mostly updates the calling convention

Its not quite settled, I don't quite like the `npy_intp *` as a default
"user-data" yet, but I think I may warm up to it, and I think some kind
of default like that is really very convenient.

Split out `user_data *`, since I think Marten has a point that
separating "constant" and volatile data is conceptionally easier.

seberg committed:
Add a bit more overview-intro at the start and refactor "bound" method section

mhvk commented on doc/neps/nep-0043-extensible-ufuncs.rst:
```diff
@@ -0,0 +1,836 @@
+
+1. For the above to work, we must also define::
+
+       np.positive.resolve_impl(MyDateTime)
```
Maybe just `MyDType` here and below? Or  in the sentence above, "1. For the above to work for, e.g., a `MyDateTime` DType, we must also define"

mhvk commented on doc/neps/nep-0043-extensible-ufuncs.rst:
```diff
@@ -0,0 +1,836 @@
+
+    res = np.positive(arr)
+
+could be implemented (conceptional) as:
```
conceptional*ly*

mhvk commented on doc/neps/nep-0043-extensible-ufuncs.rst:
```diff
@@ -0,0 +1,836 @@
+
+    add_impl = np.add.resolve_impl(type(arr1.dtype), type(arr2.dtype))
+
+This NEP strives to define the minimal set of API necessary to achieve
```
delete "set of" (i.e., just "minimal API")

mhvk commented on doc/neps/nep-0043-extensible-ufuncs.rst:
```diff
@@ -0,0 +1,836 @@
+
+   This is described in the "Promotion and Dispatching" section and
+   step 2 in the list included in the next section
+   (The Steps involved in a UFunc call).
```
Maybe just link "next section" in rst?

mhvk reviewed (commented):
@seberg - I like your example up front! Just a few super-nitpicky comments as I was reading through.

hameerabbasi reviewed (commented):
I've only read this partway through as of yet.

seberg committed:
Small fixups based on Marten and also Hameers comments

seberg commented on doc/neps/nep-0043-extensible-ufuncs.rst:
```diff
@@ -0,0 +1,631 @@
+
+4. Preparing the actual iteration. This step is largely handled by ``NpyIter`` (the iterator).
+
+   * Allocate all ouputs and potentially copies (or buffers).
```
I deleted it. For example its the implementation for overlap protection and also for very small arrays more efficient (dunno if we use it for anything else but scalars).

But: It just doesn't matter here.

seberg committed:
Revise the "promotion and dispatching" section a bit.

seberg committed:
Fix back-compat section and add a tiny bit of implementation

hameerabbasi commented on doc/neps/nep-0043-extensible-ufuncs.rst:
```diff
@@ -0,0 +1,895 @@
+
+    positive_impl = np.positive.resolve_impl(type(arr.dtype))
+
+Although, this ``positive_impl`` and ``resolve_impl`` is required for certain
+functionality, the following code::
+
+    res = positive_impl(arr)
+
+is not part of this NEP, as it is not central to the proposal.
```
Possibly move to a non-goals section.

hameerabbasi commented on doc/neps/nep-0043-extensible-ufuncs.rst:
```diff
@@ -0,0 +1,895 @@
+
+    res = np.positive(arr)
+
+could be implemented (conceptionally) as:
+
+    positive_impl = arr.dtype.positive
+    res = positive_impl(arr)
```
If this isn't what happens, and not something that the NEP will modify (as you say below), then I'd remove it completely or replace it by a more accurate statement.

seberg replied:
@hameerabbasi, I am trying to give the reader an idea/concept of what a multi-method is (and how it comes to into play with ufuncs). Starting with normal methods, seemed like a fair way to do that to me?  Maybe the sentence afterwards can say that this is not how it can be done in NumPy for reasons 1 and 2.  (Although, I have only semi appetite to discuss the possibility of an `__array_ufunc__` like approach, unless someone insists on it being listed as a rejected alternative.)

hameerabbasi replied:
How about the following:

```python
# find_actual_implementation is what we want to achieve for custom dtypes
# and ufuncs
positive_impl = find_actual_implementation(np.positive, arr.dtype)
res = positive_impl(arr)
```

I just don't want to give users who gloss over this the perception that `positive` is a `dtype` slot. It's best to choose an example which doesn't require you to explain that.

seberg replied:
Thanks, I wanted the concept of "method" in there I guess, but maybe that is just not necessary.

hameerabbasi commented on doc/neps/nep-0043-extensible-ufuncs.rst:
```diff
@@ -0,0 +1,895 @@
+
+    res = np.positive(arr)
+
+could be implemented (conceptionally) as:
```
```suggestion
could be implemented (conceptually) as:
```

hameerabbasi commented on doc/neps/nep-0043-extensible-ufuncs.rst:
```diff
@@ -0,0 +1,895 @@
+UFuncImpl Registration
+""""""""""""""""""""""
+
+*TODO:* we need to briefly mention registration, even if the details of
+how to register it are in the specs or even later!
```
Yes, we should. API decisions should come before implementation details. 😄

hameerabbasi reviewed (commented):
In general, I'd front-load the use-cases for this NEP instead of front-loading the implementation. I realize that separating the two is hard; and also that as developers who implement things, we tend to focus more on implementation.

However, this makes the end-user zone out if they don't understand, and we lose their interest. [See this discussion for details](https://mail.python.org/pipermail/numpy-discussion/2020-August/080942.html).

seberg committed:
WIP: Further updates nep 43

seberg committed:
Revise NEP 43 and expand on arraymethod (previously the dispatching/promotion was more fleshed out)

seberg committed:
Update image and make sure it is included at the right place.

seberg committed:
small fixups

dhirschfeld commented on doc/neps/nep-0043-extensible-ufuncs.rst:
```diff
@@ -0,0 +1,1261 @@
+
+The previous NEP 42 document proposes the creation of new DTypes which can
+be defined by users outside of NumPy itself.
+While implementing NEP 42 will users to create arrays with a custom dtype
```
```suggestion
While implementing NEP 42 will enable users to create arrays with a custom dtype
```

mattip commented on doc/neps/nep-0043-extensible-ufuncs.rst:
```diff
@@ -0,0 +1,1261 @@
+            # data: Pointers to the array data.
+            # strides: strides to iterate all elements
+            num_chars1 = context.descriptors[0].itemsize
+            num_chars2 = context.descriptors[0].itemsize
```
I think you need dimension as well as strides. If dimensions are in the context, then strides could be as well. Also `# n: Number of` seems like a fragment.

mattip reviewed (commented):
Here is a link to the renedered document https://766-5652112-gh.circle-artifacts.com/0/doc/neps/_build/html/nep-0043-extensible-ufuncs.html. 

I think we should merge this soon, there is an open typo suggestion and I made a suggestion around `strided_loop`, but feel free to ignore and we can iterate after this is merged.

seberg committed:
Update doc/neps/nep-0043-extensible-ufuncs.rst

Co-authored-by: Dave Hirschfeld <dave.hirschfeld@gmail.com>

seberg committed:
Fix comment which was missing the final part of sentence

seberg committed:
Pass through, up to "steps involved in a ufunc call"

seberg committed:
Review up to ArrayMethod

seberg committed:
Review up to ArrayMethod Specs

rossbar committed:
Wording updates and minor fixups.

seberg committed:
Review up to loop reuse

seberg committed:
Review most of the remaining document

seberg committed:
Fix title typo

rossbar reviewed (commented):
This is obviously quite the undertaking - thanks @seberg for taking the time to compile all this information, and for taking a lot of care in the communication of the concepts!

The only high-level comment that I have is that I would've found it useful if the "steps in a ufunc call" were higher up in the document, as this provides a nice scaffolding for all of the subsequent pieces that are addressed individually. It's not a show-stopper as there are forward references to that section in some places.

seberg committed:
Apply suggestions from code review

Co-authored-by: Ross Barnowski <rossbar@berkeley.edu>

seberg committed:
Address Ross's additional comments

seberg commented:
Thanks for the review especially Matti, Ross and of course Marten! Going to put this in so that we have an easy to read draft, but of course this is will not be the last iteration.

guidocalvano commented:
What is the state of this? The draft of this is now in master, but it is not considered an accepted NEP yet?

https://numpy.org/neps/nep-0043-extensible-ufuncs.html

seberg commented:
I am not sure we ever marked it as accepted (which we should do with the note below, probably).

The API probably moved a bit since then, but yes, I would like most of this to be available in NumPy 2.0.  We are using the API internally quite a bit.  I will note that there is still a chance of doing some changes, or omitting some functions to allow easier changes in the near future.

I am still considering one larger change mainly:  Re-adding the static `void *` pointer as part of the `context` struct as well as moving the new `auxdata` into the `context` struct.  If you have any thoughts, please let us know, this is the last few weeks were changes are trivial.
(The new API is designed to be able to deprecate older versions, but especially for the ufunc inner-loop signature that seems tedious.)

The new API is available for experimentation for a long time via `extermintal_dtype_api.h`.

guidocalvano commented:
Is there a python api already? I don't expect to need c code.

Some more context should you be interested:
What I want to do is implement two things: decimal types and multivectors from geometric algebra. I expect both to be implementable fairly easily and quickly by a python ufunc: i.e. you create a dtype that imposes a size on an element, and then you can create a view of a different dtype, to access the data. Then you can do your math by operating on the view and return whatever the operator requires. Decimal add can be done by simply viewing the values as ints and adding them. Multiplication can be done by promoting to a bigger int data type, doing the multiplication, then shifting values two times the numbers behind the decimal point. Multivectors are a bit more complicated... it would take at least half an hour to explain the concept... but well worth watching a few videos on.

seberg commented:
I have no plans for Python API for the foreseeable future, it would need a higher level hook to do well probably.  Yes, you could add that, but no it doesn't exist.

mhvk commented:
@seberg - I think an ability to pass on arbitrary information to the inner loop remains a good idea. E.g., using old-style I have used it to create "ufunc chains" where the inner loop executes other inner loops in order - for speed-ups comparable to numexpr without having to describe operations in strings (see https://github.com/mhvk/chain_ufunc). And of course, you saw that it was somewhat handy for the gufunc fft/ifft routines in #25536...

But the main problem in commenting is that while I reviewed quite a bit, I haven't yet *used* the new API, so I don't really know how it would work in practice... So take this with a pinch of salt!
